fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/cursors/throttled/ThrottledIteratorTest.java

```diff
@@ -0,0 +1,588 @@
+                simpleRec.mergeFrom(rec.getRecord());
+                itemsScanned.add((int)simpleRec.getRecNo());
+                return null;
+            });
+        };
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            for (int i = 0; i < numRecords; i++) {
+                final TestRecords1Proto.MySimpleRecord record = TestRecords1Proto.MySimpleRecord.newBuilder()
+                        .setRecNo(i)
+                        .setStrValueIndexed("Some text")
+                        .setNumValue3Indexed(1415 + i * 7)
+                        .build();
+                recordStore.saveRecord(record);
+            }
+            commit(context);
+        }
+
+        // For this test, start and finalize the iteration within the transaction
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            ThrottledRetryingIterator.Builder<Tuple> builder = ThrottledRetryingIterator
+                    .builder(fdb, cursorFactory, itemHandler)
+                    .withNumOfRetries(2)
+                    .withMaxRecordsScannedPerTransaction(maxRowLimit);
+            try (ThrottledRetryingIterator<Tuple> iterator = builder.build()) {
+                iterator.iterateAll(recordStore.asBuilder()).join();
+            }
+        }
+        assertThat(itemsScanned).isEqualTo(IntStream.range(0, numRecords).boxed().collect(Collectors.toList()));
+
+        // For this test, start iteration within the transaction but allow it to run (and create more transactions) outside
+        // of the original transaction
+        itemsScanned.clear();
+        ThrottledRetryingIterator<Tuple> iterator = ThrottledRetryingIterator
+                .builder(fdb, cursorFactory, itemHandler)
+                .withNumOfRetries(2)
+                .withMaxRecordsScannedPerTransaction(maxRowLimit)
+                .build();
+        CompletableFuture<Void> iterateAll;
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            iterateAll = iterator.iterateAll(recordStore.asBuilder());
+        }
+        iterateAll.join();
+        iterator.close();
+        assertThat(itemsScanned).isEqualTo(IntStream.range(0, numRecords).boxed().collect(Collectors.toList()));
+    }
+
+    @Test
+    void testLateCompleteFutures() throws Exception {
+        int numRecords = 50;
+        List<CompletableFuture<Void>> futures = new ArrayList<>(numRecords);
+
+        final ItemHandler<Integer> itemHandler = (store, item, quotaManager) -> {
+            // First future hangs on, all others are immediately completed
+            CompletableFuture<Void> future = (item.get() == 0) ? new CompletableFuture<>() : CompletableFuture.completedFuture(null);
+            futures.add(future);
+            return future;
+        };
+
+        ThrottledRetryingIterator<Integer> throttledIterator =
```

Comment:
Should this test, and the one below use `try-with-resources`

Reply:
I wanted these tests, since they were testing anomalous closing scenarios, to have explicit `close` calls for clarity.
This also changes things up a little for the runner as there are things happening *after* the transaction closed.